jerry-core/include/jerryscript-compiler.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYSCRIPT_COMPILER_H
+#define JERRYSCRIPT_COMPILER_H
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif /* __cplusplus */
+
+/** \addtogroup jerry-compiler Jerry compiler compatibility components
+ * @{
+ */
+
+/*
+ * Attributes
+ */
+#define JERRY_ATTR_ALIGNED(ALIGNMENT) __attribute__((aligned(ALIGNMENT)))
+#define JERRY_ATTR_ALWAYS_INLINE __attribute__((always_inline))
+#define JERRY_ATTR_CONST __attribute__((const))
+#define JERRY_ATTR_DEPRECATED __attribute__((deprecated))
+#define JERRY_ATTR_FORMAT(...) __attribute__((format(__VA_ARGS__)))
+#define JERRY_ATTR_HOT __attribute__((hot))
+#define JERRY_ATTR_NOINLINE __attribute__((noinline))
+#define JERRY_ATTR_NORETURN __attribute__((noreturn))
+#define JERRY_ATTR_PURE __attribute__((pure))
+#define JERRY_ATTR_SECTION(SECTION) __attribute__((section(SECTION)))
+#define JERRY_ATTR_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
+
+/*
+ * Condition likeliness, unlikeliness
+ */
+#define JERRY_LIKELY(x) __builtin_expect(!!(x), 1)
+#define JERRY_UNLIKELY(x) __builtin_expect(!!(x), 0)
```

Comment:
I like these names. But my question still stands: what happens on non-GCC systems? (Btw clang supports GNU macro)

Reply:
MSVC doesn't support, define it to empty.

Reply:
Which will (have to) happen in #2306

Reply:
@zherczeg My previous answer is hopefully still an answer:
> So, for now at least, I deliberately dropped the macro guards as gcc and clang are the officially maintained and tested compilers (Travis CI is green!), accepting that we may see some follow-up patches from hopefully-not-too-upset users of different compilers. But this time, we will hopefully have more precise guards and only where they are needed and all in one place.

So, I'm happy to take the blame for some fallout and review the upcoming PRs that add support for non-gcc and non-clang compatible compilers.

Reply:
Well, OK, let's think ahead a bit. Should the macros in `-compiler.h` be opt-in or opt-out?

Opt-in:
```c
/* Compiler-specific macros relevant for GCC. */
#ifdef __GNUC__
#define JERRY_ATTR_ALIGNED(ALIGNMENT) __attribute__((aligned(ALIGNMENT)))
// snip...
#define JERRY_LIKELY(x) __builtin_expect(!!(x), 1)
// snip...
#endif

/* Default empty definitions for all compiler-specific macros. Define any of
 * these in a guarded block above (e.g., as for GCC) to fine tune compilation
 * for your own compiler. */
#ifndef JERRY_ATTR_ALIGNED
#define JERRY_ATTR_ALIGNED(ALIGNMENT)
#endif

#ifndef JERRY_LIKELY
#define JERRY_LIKELY(x)
#endif
```
This is safe as all macros default to the empty string, so all compilers are supported by default. But the default will be sub-optimal then.

Opt-out:
```c
/* Default definitions for compiler-specific macros, whatever is supported by
 * GCC. If your compiler doesn't support any of these or uses different
 * notation, redefine them in a guarded block below. */
#define JERRY_ATTR_ALIGNED(ALIGNMENT) __attribute__((aligned(ALIGNMENT)))
// snip...
#define JERRY_LIKELY(x) __builtin_expect(!!(x), 1)
// snip...

/* Compiler-specific macros for MSVC. Define almost everything as empty because
 * of the differences in the __attribute__ and __declspec syntax. */
#ifdef _MSC_VER
#undef JERRY_ATTR_ALIGNED
#define JERRY_ATTR_ALIGNED(ALIGNMENT)

#undef JERRY_LIKELY
#define JERRY_LIKELY(x)
#endif
```
This is optimal for the officially supported/tested compilers (GCC and Clang). But it needs more maintenance if anything changes (e.g., all compiler blocks have to be adapted if a new attribute gets used, otherwise we risk breaking them).

Or will the reality become much more complex than opt-on or opt-out?

Reply:
I like the Opt-in one:)